secret-manager/pkg/backend/cache/backend.go

```diff
@@ -41,36 +43,55 @@ func (c *CachedBackend[T, S]) ParseSecretId(raw string) (T, error) {
 
 func (c *CachedBackend[T, S]) Get(ctx context.Context, id T) (res S, err error) {
 	log := logr.FromContextOrDiscard(ctx)
-	if item, ok := c.Cache.Get(id.String()); ok && !item.Expired() {
-		metrics.RecordCacheHit()
-		return item.Value(), nil
+	cacheId := id.Copy().(T)
+	cacheKey := cacheId.String()
+
+	if item, ok := c.Cache.Get(cacheKey); ok && !item.Expired() {
+		metrics.RecordCacheHit("")
+		return item.Value().Copy().(S), nil
 	}
 
 	metrics.RecordCacheMiss("not_found")
-	log.Info("Cache miss", "id", id.String())
-	item, err := c.Backend.Get(ctx, id)
+	item, err := c.Backend.Get(ctx, cacheId)
 	if err != nil {
 		return res, err
 	}
+	if item.Value() == "" {
+		// Do not cache empty secrets
+		return item, nil
+	}
+	copy := item.Copy().(S)
+
+	log.V(1).Info("Caching Get result", "id", copy.Id().String())
+	c.Cache.Set(cacheKey, NewDefaultCacheItem(copy.Id(), copy, c.ttl))
 
-	c.Cache.Set(id.String(), NewDefaultCacheItem(id, item, c.ttl))
 	return item, nil
 }
 
 func (c *CachedBackend[T, S]) Set(ctx context.Context, id T, value backend.SecretValue) (res S, err error) {
-	if item, ok := c.Cache.Get(id.String()); ok {
-		if value.EqualString(item.Value().Value()) {
-			return item.Value(), nil
+	cacheId := id.Copy()
+
+	if item, ok := c.Cache.Get(cacheId.String()); ok && !item.Expired() {
+		cachedSecret := item.Value()
+		if value.EqualString(cachedSecret.Value()) {
+			metrics.RecordCacheHit("set")
+			return cachedSecret.Copy().(S), nil
+		} else {
+			metrics.RecordCacheMiss("value_mismatch")
+			c.Cache.Delete(cacheId.String())
 		}
 	}
-	item, err := c.Backend.Set(ctx, id, value)
+
+	metrics.RecordCacheMiss("set")
```

Comment:
Similar to my other comment regarding the metrics for the cache.

Here, I propose: `cache_access_total{result="miss", method="set", reason: "not_found"}`